x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2857 @@
+        <td><code>Prebuilt rule name</code></td>
+        <td><code>Completely different string</code></td>
+        <td style="border-right:3px solid black"><code>Updated prebuilt rule name</code></td>
+        <td><code>Completely different string</code></td>
+      </tr>
+      <tr>
+        <td style="border-right:3px solid black">Keep user-added prefixes and accept other modifications</td>
+        <td><code>Prebuilt rule name</code></td>
+        <td><code>[Security Solution] Prebuilt rule name</code></td>
+        <td style="border-right:3px solid black"><code>Updated prebuilt rule name</code></td>
+        <td><code>[Security Solution] Updated prebuilt rule name</code></td>
+      </tr>
+      <tr>
+        <td style="border-right:3px solid black">Keep user-added suffixes and accept other modifications</td>
+        <td><code>Prebuilt rule name</code></td>
+        <td><code>Prebuilt rule name - (Custom version 3)</code></td>
+        <td style="border-right:3px solid black"><code>Updated prebuilt rule name</code></td>
+        <td><code>Updated prebuilt rule name - (Custom version 3)</code></td>
+      </tr>
+      <tr>
+        <td style="border-right:3px solid black">Keep user customizations on conflicts, but add updates in sections with no conflicts</td>
+        <td><code>Prebuilt rule name</code></td>
+        <td><code>Customized prebuilt rule name</code></td>
+        <td style="border-right:3px solid black"><code>Updated prebuilt rule name. Much better.</code></td>
+        <td><code>Customized prebuilt rule name. Much better.</code></td>
+      </tr>
+    </tbody>
+  </table>
+
+
+##### Number fields
+
+Number fields should be treated as a whole unit, i.e. not breakable by digits. Therefore, there is only one possibility of conflicts, the scenario (A B C). In that case, **keep the current version** with the user customization.
+
+Examples: `risk_score`, `max_signals`
+
+  <table>
+    <thead>
+      <tr>
+        <th style="border-right:3px solid black">Use case</th>
+        <th>Base version</th>
+        <th>Current version</th>
+        <th style="border-right:3px solid black">Target version</th>
+        <th>Merged version (output)</th>
+      </tr>
+    </thead>
+    <tbody align="center">
+      <tr>
+        <td style="border-right:3px solid black">Keep customizations if conflict</td>
+        <td><code>50</code></td>
+        <td><code>65</code></td>
+        <td style="border-right:3px solid black"><code>20</code></td>
+        <td><code>65</code></td>
+      </tr>
+    </tbody>
+  </table>
+
+##### Array fields
+
+###### Array of strings fields
+
+For **array of strings** fields, conflict resolution will take place on an element by element basis - that means that the algorithm will not consider or try to solve changes within a single element. For example:
+
+```
+base: [test1]
+current: [my-test1]
+target: [test2]
+output: [my-test1]
```

Comment:
I don't understand the logic behind this example. Can you elaborate what is it and how that algorithm should work (at a very high level)?

My expectation would be the following:

```
base: [linux, network]
current: [windows, host]
target: [linux, ml]
output: [windows, host, ml]
```

The logic is a bunch of simple set operations:

- Changes to `current` compared against `base`:
  - Added: [`windows`, `host`]
  - Removed: [`linux`, `network`]
- Changes to `target` compared against `base`:
  - Added: [`ml`]
  - Removed: [`network`]
- In combination:
  - Added: [`windows`, `host`, `ml`]
  - Removed: [`linux`, `network`]
- Which, after applying to `base` ([`linux`, `network`]), gives the result [`windows`, `host`, `ml`]

Reply:
Fixed this whole section.

@banderror @maximpn 
I do have one disagreement with this logic though, and would like to ask you for your opinion.

In the example given above, the user removed [linux, network] and added [windows, host], while the Elastic updated added [ml] and removed [network].

I agree that the end result should include all the added elements in both modifications so:
[windows, host] + [ml] -> [windows, host, ml] should be added.

But I think that the logic for removals should probably be different. I see two possibilities:
1. Remove the element if it was removed in BOTH operations. This would mean:
[linux, network] + [network] -> only [network] should be removed (and linux shouldn't). So basically, an **intersection** of both removal operations.

2. Remove the element if it was removed in the current version. The UX logic behind this is that we probably don't want to remove an index patter (for example) if the user has not expressly deleted it from their rule, just because the Elastic update proposes it. That might break the rule. In this case, it would work so:
```js
const base = new Set(["linux", "network"]);  
const current = new Set(["linux", "ml", "security]);  
const target = new Set(["aws", "cloudfront"]);  

const addedCurrent = current.difference(base); // [ml, security]
const removedCurrent = base.difference(current); // [network]

const addedTarget = target.difference(base); // [aws, cloudfront]

const merged = base
                .difference(removedCurrent)
                .union(addedCurrent)
                .union(addedTarget)

// Results in: [linux, ml, security, aws, cloudfront]
```
I think this is a more conservative approach, less-riskier for users and less prone to accidental broken rules.

What do you think? The RFC currently has option 1, but I think I'd rather change the approach to option 2.

Reply:
@jpdjere I'm afraid code example you provided in the second case isn't fully reflecting the proposal. Both `current` and `target` versions don't have `network` but you say `Remove the element if it was removed in the current version. Or I just didn't what you meant here.

Reading explanations we have two options

1. remove an element if it was removed in `current` and `target` versions
2. remove an element if it was removed only in the `current` version

I'd say if user removed something it shouldn't appear again. So the second case makes sense.

Reply:
@jpdjere @maximpn Here are my thoughts on that:

> Remove the element if it was removed in BOTH operations.
> remove an element if it was removed in `current` and `target` versions

Why this would be needed? If I, as a user, decided to remove a certain tag from an array of tags (maybe I decided to completely replace the prebuilt tagging system with my own custom set of tags), and the auto-merge algorithm adds it back for me just because Elastic didn't remove this tag, wouldn't this annoy me?

From a different angle: why my user customizations (deletions, in this case) MUST be echoed by updates from Elastic in order to be taken into account during rule upgrade?

From another different angle: if Elastic removed a tag (because, for example, Elastic decided to rework the prebuilt tagging system), why this change MUST be echoed by the same user customization in order to be taken into account during rule upgrade?

> Remove the element if it was removed in the current version.
> remove an element if it was removed only in the `current` version
> The UX logic behind this is that we probably don't want to remove an index pattern (for example) if the user has not expressly deleted it from their rule, just because the Elastic update proposes it. That might break the rule.

Why do you think updates from Elastic should be "weaker" than user customizations, when it comes to removing items from arrays? What if as part of fixing or tuning a rule Elastic needs to remove a certain index pattern, if it was previously added by mistake? Or replace a broader index pattern by a narrower index pattern (which would be == a removal + an addition)? Or, as mentioned above, what if Elastic wants to rework the set of prebuilt tags, and replace most of existing tags with new ones? Prebuilt tags were completely reworked a few months ago btw.

> const base = new Set(["linux", "network"]);  
> const current = new Set(["linux", "ml", "security"]);  
> const target = new Set(["aws", "cloudfront"]);  
> Results in: [linux, ml, security, aws, cloudfront]

Isn't it the same as `current.union(target)`? Not sure why the proposed code is more complicated than that.

----------

Considering that I'd stick to my proposal.

@approksiu Can you please review this thread of comments and share your opinion? This is something that can have an impact on UX.

Reply:
From yet another angle:

> remove an element if it was removed in BOTH current and target versions
> remove an element if it was removed only in the current version

Just an example, what would happen if the diff algo would use either if these two heuristics:

```
base: [logs-*]
current: [logs-*]
target: [logs-endpoint.events-*]
output: [logs-*, logs-endpoint.events-*]
```

Reply:
- Revert back to first algorithm, but update Conflict to `SOLVABLE`